File: 0001-two-sum/0001-two-sum.java
class Solution {
    public int[] twoSum(int[] nums, int target) {
        int [][] a = new int[nums.length][2];

        for (int i = 0; i < nums.length; i++) {
            a[i][0] = nums[i]; // value
            a[i][1] = i;       // original index
        }

        Arrays.sort(a, (x, y) -> Integer.compare(x[0], y[0]));

        int l = 0, r = a.length - 1;

        while (l < r) {
            int sum = a[l][0] + a[r][0];

            if (sum == target)
                return new int[]{a[l][1], a[r][1]};
            else if (sum < target)
                l++;
            else
                r--;
        }

        return new int[]{};
    }
}

File: 0001-two-sum/0001-two-sum.cpp
// class Solution {
// public:
//     vector<int> twoSum(vector<int>& nums, int target) {
//         unordered_map<int, int> seen;
//         for (int i = 0; i < nums.size(); ++i) {
//             int complement = target - nums[i]; 
//             if (seen.find(complement) != seen.end()) {
//                 return {seen[complement], i};
//             }
//             seen[nums[i]] = i;
//         }
//         return {};
//     }

// //-----Brute force-----
//     //  int n = nums.size();        
//     //     for (int i = 0; i < n; ++i) {
//     //         for (int j = i + 1; j < n; ++j) {
//     //             if (nums[i] + nums[j] == target) {
//     //                 return {i, j};
//     //             }
//     //         }
//     //     }
//     //     return {};
//     // }



// };

class Solution {
public:
    vector<int> twoSum(vector<int>& nums, int target) {
        vector<pair<int,int>> a;

        for(int i = 0; i < nums.size(); i++)
            a.push_back({nums[i], i});

        sort(a.begin(), a.end());

        int l = 0, r = a.size() - 1;

        while(l < r) {
            int sum = a[l].first + a[r].first;

            if(sum == target)
                return {a[l].second, a[r].second};
            else if(sum < target)
                l++;
            else
                r--;
        }

        return {};
    }
};